### ISSUE TITLE
Flux sync issue - err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository

### ISSUE THREAD

**Issue by nikhilbalekundargi** [fluxcd/flux#3042]
**Describe the bug**

Flux sync --k8s-fwd-ns=flux not working.

Error: git repository ssh://git@github.com/XXX-YYY/ZZZ-123 is not ready to sync (status: unreachable)
Run 'fluxctl sync --help' for usage.

Below are the flux logs -
component=sync-loop err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository., full output:\n Cloning into bare repository '/tmp/flux-gitclon'...\nssh: Could not resolve hostname github.com: Try again\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
Flux version - 1.19.0 (edited

**To Reproduce**

Steps to reproduce the behaviour:

1. Provide Flux install instructions
2. Provide a GitHub repository with Kubernetes manifests

**Expected behavior**
Flux sync should work and flux should sync periodically

**Logs**
ts=2020-05-04T11:39:43.248157504Z caller=images.go:27 component=sync-loop msg="no automated workloads"
ts=2020-05-04T11:41:24.77104663Z caller=loop.go:107 component=sync-loop err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository., full output:\n Cloning into bare repository '/tmp/flux-gitclone606183439'...\nssh: Could not resolve hostname github.com: Try again\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
ts=2020-05-04T11:44:43.248314115Z caller=images.go:17 component=sync-loop msg="polling for new images for automated workloads"
ts=2020-05-04T11:44:43.248377503Z caller=images.go:27 component=sync-loop msg="no automated workloads"
ts=2020-05-04T11:46:24.771300069Z caller=loop.go:107 component=sync-loop err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository., full output:\n Cloning into bare repository '/tmp/flux-gitclone606183439'...\nssh: Could not resolve hostname github.com: Try again\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
ts=2020-05-04T11:49:43.248602614Z caller=images.go:17 component=sync-loop msg="polling for new images for automated workloads"
ts=2020-05-04T11:49:43.249028249Z caller=images.go:27 component=sync-loop msg="no automated workloads"
ts=2020-05-04T11:51:24.771622319Z caller=loop.go:107 component=sync-loop err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository., full output:\n Cloning into bare repository '/tmp/flux-gitclone606183439'...\nssh: Could not resolve hostname github.com: Try again\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
ts=2020-05-04T11:54:43.24931619Z caller=images.go:17 component=sync-loop msg="polling for new images for automated workloads"
ts=2020-05-04T11:54:43.249473129Z caller=images.go:27 component=sync-loop msg="no automated workloads"
ts=2020-05-04T11:56:24.772004282Z caller=loop.go:107 component=sync-loop err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository., full output:\n Cloning into bare repository '/tmp/flux-gitclone606183439'...\nssh: Could not resolve hostname github.com: Try again\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"
ts=2020-05-04T11:59:43.249723538Z caller=images.go:17 component=sync-loop msg="polling for new images for automated workloads"
ts=2020-05-04T11:59:43.250083855Z caller=images.go:27 component=sync-loop msg="no automated workloads"
ts=2020-05-04T12:01:24.772483917Z caller=loop.go:107 component=sync-loop err="git repo not ready: git clone --mirror: fatal: Could not read from remote repository., full output:\n Cloning into bare repository '/tmp/flux-gitclone606183439'...\nssh: Could not resolve hostname github.com: Try again\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"

**Additional context**

- Flux version:  1.19.0
- Kubernetes version:
Client Version: v1.15.3
Server Version: v1.14.10-gke.27
- Git provider: Github
- Container registry provider: GCR

**Comment by ambis** [comment]
I have this exact same problem on personal VPS k3s installs as well as corp. GKE installs.

When exec'ing to the pod, the pod resolves gitlab/github properly (eg. `getent ahosts gitlab.com`), but the flux process maybe caches the IP for too long and does not re-resolve?

**Comment by nikhilbalekundargi** [comment]
> I have this exact same problem on personal VPS k3s installs as well as corp. GKE installs. Flux `1.19.0`.
> 
> When exec'ing to the pod, the pod resolves gitlab/github properly (eg. `getent ahosts gitlab.com`), but the flux process maybe caches the IP for too long and does not re-resolve?
> 
> Deleting the pod helps as it re-resolves the name.

Yes, restarting the pod helped me.

**Comment by ambis** [comment]
I counted that I'm working with 8 instances of flux 1.19.0 daily. Chances that one or multiple of them are in this state is very high. And I mean high in the sense of about 2-3 flux pod restarts needed per day. 

Well, one of them talks to our internal gitlab host, so that doesn't fail because the gitlab host IP never changes.

**Comment by bpinter** [comment]
Hello. I have the same issue.

**Comment by tshak** [comment]
I've experienced the same symptoms on two different AKS clusters. Restarting the pod resolves the issue.

Flux: `1.19.0`
Kube Server: `v1.16.9`

**Comment by brokenjacobs** [comment]
Seems to be 1.19 specific, or at least worse on 1.19.

**Comment by 2mf** [comment]
I can also confirm this issue.
We have several AKS Cluster with flux 1.19 and have observed this issue 1 or twice a week.

**Comment by seeruk** [comment]
Also seeing this on GKE, K8s 1.16, also version 1.19.0.

**Comment by terrycain** [comment]
Am seeing this with Flux 1.19.0 on Kubernetes 1.15 and 1.17

@hiddeco Any idea on what would cause this behaviour or what could be done to troubleshoot it further?

**Comment by hiddeco** [comment]
This is a duplicate of #3011 and has been fixed in #3013.

**Comment by terrycain** [comment]
Thanks @hiddeco

**Comment by 2mf** [comment]
When will there be a new release, since I can't not justify using a nightly build or RC for production environments.

**Comment by metost** [comment]
> When will there be a new release, since I can't not justify using a nightly build or RC for production environments.

https://github.com/fluxcd/flux/releases/tag/1.20.0

**Comment by andyfcx** [comment]
Using flux 1.20.2, having the same problem

**Comment by ambis** [comment]
> Using flux 1.20.2, having the same problem

After the initial fix, none of the ~7 flux instances I run in different envs has ever failed with this again. 

Just saying that the initial fix did fix it for me.

**Comment by kingdonb** [comment]
This was fixed in 1.20.0, thanks for reporting!

I can confirm, I have been keeping an instance of Flux on 1.19.x for some time past its expiration date (don't ask why... ok, it's because of okteto/charts#109) and it has suffered this problem the whole time, until recently when I upgraded it.

Appears to be solved with confirmation from a maintainer about [the source of the fix](https://github.com/fluxcd/flux/issues/3042#issuecomment-648059485) and it has landed in a release, so closing.
